pgex/examples/collision_shapes/main.py

```diff
@@ -0,0 +1,335 @@
+    radii = (circle_a.radius + circle_b.radius) ** 2
+
+    return distance <= radii
+
+def rect_x_circle(rect: pygame.FRect, circle: Circle) -> bool:
+    """ Check collision between rectangle and circle. """
+
+    center = circle.center.copy()
+
+    # Rect borders
+    right = rect.x + rect.width
+    left = rect.x
+    bottom = rect.y + rect.height
+    top = rect.y
+
+    # Check circle center against rect borders
+    if center.x > right: center.x = right
+    elif center.x < left: center.x = left
+    if center.y > bottom: center.y = bottom
+    elif center.y < top: center.y = top
+
+    # Check distance against original center
+    return center.distance_squared_to(circle.center) < circle.radius ** 2
+
+def project_circle(
+        circle: Circle,
+        axis: pygame.Vector2
+        ) -> tuple[float, float]:
+    """ SAT helper function to project circle onto an axis. """
+
+    axis_radius = axis.normalize() * circle.radius
+
+    p1 = circle.center + axis_radius
+    p2 = circle.center - axis_radius
+
+    min_ = p1.dot(axis)
+    max_ = p2.dot(axis)
+
+    # Swap min max values if needed
+    if (min_ > max_):
+        min_, max_ = max_, min_
+
+    return (min_, max_)
+
+def project_polygon(
+        polygon: Polygon,
+        axis: pygame.Vector2
+        ) -> tuple[float, float]:
+    """ SAT helper function to project polygon vertices onto an axis. """
+
+    min_ = float("inf")
+    max_ = float("-inf")
+
+    for vertex in polygon.local_to_world():
+        projection = vertex.dot(axis)
+
+        if projection < min_: min_ = projection
+
+        if projection > max_: max_ = projection
+
+    return (min_, max_)
+
+def polygon_x_circle(polygon: Polygon, circle: Circle) -> bool:
+    """ Check collision between polygon and circle. """
+
+    vertices = polygon.local_to_world()
+    n = len(vertices)
+
+    for i in range(n):
+        va = vertices[i]
+        vb = vertices[(i + 1) % n]
+
+        edge = vb - va
+        axis = edge.normalize()
+        axis = pygame.Vector2(-axis.y, axis.x)
+
+        min_a, max_a = project_polygon(polygon, axis)
+        min_b, max_b = project_circle(circle, axis)
+
+        if min_a >= max_b or min_b >= max_a:
+            return False
+    
+    return True
+
+def polygon_x_rect(polygon: Polygon, rect: pygame.FRect) -> bool:
+    """ Check collision between polygon and rectangle. """
+
+    # IMPORTANT: A better approach would be clipping the polygon and testing
+    #            against the rect. But for now we are just treating the rect
+    #            as another polygon.
+
+    rect_center = pygame.Vector2(*rect.center)
+
+    rect_vertices = [
+        pygame.Vector2(*rect.topleft) - rect_center,
+        pygame.Vector2(*rect.topright) - rect_center,
+        pygame.Vector2(*rect.bottomright) - rect_center,
+        pygame.Vector2(*rect.bottomleft) - rect_center
+    ]
+
+    rect_poly = Polygon(rect_center, rect_vertices)
```

Comment:
Since you are using the variable in only one place and it's right below, I guess it could be inlined.